### ISSUE TITLE
.Net 8 - `The type 'Postgrest.Attributes.PrimaryKeyAttribute' is not a supported dictionary key using converter of type 'System.Text.Json.Serialization.Converters.SmallObjectWithParameterizedConstructorConverter`

### ISSUE THREAD

**Issue by jimweigandt** [supabase-community/supabase-csharp#174]
# Bug report

## Describe the bug

I realize this is the same exact bug [that was reported in the past for .Net 7](https://github.com/supabase-community/supabase-csharp/issues/72) but I am running into the same type error in .Net 8. [The suggested fix is also not working for me](https://github.com/supabase-community/supabase-csharp/issues/72#issuecomment-1562237680). But specifically I am trying to add a new record to my Supabase instance but it's tripping up when using the `PrimaryKey` attribute. I can do GET requests just fine, however.

It seems I still can't use the `AddNewtonsofJson()` method to override using `System.Text.JSON` as the serializer/deserialize. Any tips on navigating this? 

Here are all the packages that I've tried to install to try and get NewtonsoftJSON working.

```csharp
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.JsonPatch" Version="8.0.10" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="8.0.10" />
    <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="8.0.8" />
    <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
    <PackageReference Include="Newtonsoft.Json.Bson" Version="1.0.2" />
    <PackageReference Include="postgrest-csharp" Version="3.5.1" />
    <PackageReference Include="supabase-csharp" Version="0.16.2" />
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.4.0" />
  </ItemGroup>
```

Here is my `Program.cs` file:

```csharp
using Supabase;
using WebApiTest.Contracts;
using WebApiTest.Models;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddControllers().AddNewtonsoftJson();

// Add services to the container.
// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

builder.Services.AddScoped<Supabase.Client>(_ =>
  new Client(
    builder.Configuration["SUPABASE_URL"],
    builder.Configuration["SUPABASE_KEY"],
    new SupabaseOptions
    {
      AutoRefreshToken = true,
      AutoConnectRealtime = true
    }));

var app = builder.Build();

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.UseHttpsRedirection();

var summaries = new[]
{
    "Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"
};

app.MapGet("/weatherforecast", () =>
{
    var forecast =  Enumerable.Range(1, 5).Select(index =>
        new WeatherForecast
        (
            DateOnly.FromDateTime(DateTime.Now.AddDays(index)),
            Random.Shared.Next(-20, 55),
            summaries[Random.Shared.Next(summaries.Length)]
        ))
        .ToArray();
    return forecast;
})
.WithName("GetWeatherForecast")
.WithOpenApi();

app.MapPost("/newsletters", async (CreateNewsletterRequest request, Client client) =>
{
  var newsletter = new Newsletter
  {
    Name = request.Name,
    Description = request.Description,
    ReadTime = request.ReadTime,
  };

  var response = await client.From<Newsletter>().Insert(newsletter);

  var newNewsletter = response.Models.First();
  return Results.Ok(newNewsletter);
});

app.MapGet("/newsletters/{id}", async (long id, Client client) =>
{
  var response = await client
    .From<Newsletter>()
    .Where(n => n.Id == id)
    .Get();
  
  var newsletter = response.Models.FirstOrDefault();

  if (newsletter is null)
  {
    return Results.NotFound();
  }

  var newsletterResponse = new NewsletterResponse
  {
    Id = newsletter.Id,
    Name = newsletter.Name,
    Description = newsletter.Description,
    ReadTime = newsletter.ReadTime,
    CreatedAt = newsletter.CreatedAt,
  };

  return Results.Ok(newsletterResponse);
});

app.MapDelete("/newsletters/{id}", async (long id, Client client) =>
{
  await client
    .From<Newsletter>()
    .Where(n => n.Id == id)
    .Delete();
  
  return Results.NoContent();
});

app.Run();

record WeatherForecast(DateOnly Date, int TemperatureC, string? Summary)
{
    public int TemperatureF => 32 + (int)(TemperatureC / 0.5556);
}
```

Here is the model I'm trying to use to send a POST request:
```csharp
using Postgrest.Attributes;
using Postgrest.Models;

namespace WebApiTest.Models
{
  [Table("newsletters")]
  public class Newsletter : BaseModel
  {
    [PrimaryKey("id")]
    public long Id { get; set; }

    [Column("name")]
    public string Name { get; set; }

    [Column("description")]
    public string Description { get; set; }

    [Column("read_time")]
    public int ReadTime { get; set; }

    [Column("created_at")]
    public DateTime CreatedAt { get; set; }
  }
}
```
Here is the full error:
```
System.NotSupportedException: The type 'Postgrest.Attributes.PrimaryKeyAttribute' is not a supported dictionary key using converter of type 'System.Text.Json.Serialization.Converters.SmallObjectWithParameterizedConstructorConverter`5[Postgrest.Attributes.PrimaryKeyAttribute,System.String,System.Boolean,System.Object,System.Object]'. Path: $.PrimaryKey.
 ---> System.NotSupportedException: The type 'Postgrest.Attributes.PrimaryKeyAttribute' is not a supported dictionary key using converter of type 'System.Text.Json.Serialization.Converters.SmallObjectWithParameterizedConstructorConverter`5[Postgrest.Attributes.PrimaryKeyAttribute,System.String,System.Boolean,System.Object,System.Object]'.
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(Type keyType, JsonConverter converter)
   at System.Text.Json.Serialization.JsonConverter`1.WriteAsPropertyName(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonConverter`1.WriteAsPropertyNameCore(Utf8JsonWriter writer, T value, JsonSerializerOptions options, Boolean isWritingExtensionDataProperty)
   at System.Text.Json.Serialization.Converters.DictionaryOfTKeyTValueConverter`3.OnWriteResume(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonDictionaryConverter`3.OnTryWrite(Utf8JsonWriter writer, TDictionary dictionary, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(WriteStack& state, NotSupportedException ex)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCoreAsObject(Utf8JsonWriter writer, Object value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.JsonSerializer.WriteCore[TValue](Utf8JsonWriter writer, TValue& value, JsonTypeInfo jsonTypeInfo, WriteStack& state)
   at System.Text.Json.JsonSerializer.WriteStreamAsync[TValue](Stream utf8Json, TValue value, JsonTypeInfo jsonTypeInfo, CancellationToken cancellationToken)
   at System.Text.Json.JsonSerializer.WriteStreamAsync[TValue](Stream utf8Json, TValue value, JsonTypeInfo jsonTypeInfo, CancellationToken cancellationToken)
   at System.Text.Json.JsonSerializer.WriteStreamAsync[TValue](Stream utf8Json, TValue value, JsonTypeInfo jsonTypeInfo, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Http.HttpResponseJsonExtensions.WriteAsJsonAsyncSlow(Stream body, Object value, Type type, JsonSerializerOptions options, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.ExecuteTaskResult[T](Task`1 task, HttpContext httpContext)
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass89_2.<<HandleRequestBodyAndCompileRequestDelegateForJson>b__2>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Go to '…'
2. Click on '…'
3. Scroll down to '…'

## Expected behavior

Send a POST request and add a new record to a table in Supabase

## System information

- OS: Mac
- Version of supabase-csharp: 0.16.2
- Version of .Net Core: 8.0

**Comment by amolwankhede** [comment]
likewise - having same issue 

System.NotSupportedException: The type 'Supabase.Postgrest.Attributes.PrimaryKeyAttribute' is not a supported dictionary key using converter of type 'System.Text.Json.Serialization.Converters.SmallObjectWithParameterizedConstructorConverter`5[Supabase.Postgrest.Attributes.PrimaryKeyAttribute,System.String,System.Boolean,System.Object,System.Object]'. Path: $.Value.PrimaryKey.

Tried adding package "Json.net" oberlik

**Comment by frasermclean** [comment]
I experienced this same issue and worked out the problem. The model you are trying to serialize `Newsletter`, extends `BaseModel` and the serializer if having issues with that. You need to map it to another response class e.g. `NewsletterResponse` that contains only the properties you want to receive.

**Comment by frasermclean** [comment]
Looking at the `BaseModel` class it has `JsonIgnore` attributes from `Newtonsoft` JSON. So if you were using `Newtonsoft` serializer it would probably work. However the default serializer for ASP.Net Core is the built in `System.Text.Json`. Not sure why the Supabase package depends on Newtonsoft JSON. It's considered pretty legacy now.

**Comment by jimweigandt** [comment]
FWIW I checked my database and apparently my POST requests _were_ working at some point despite that error returning on every request. My guess is it probably had something to do with setting my endpoints directly into `Program.cs` and when I added `AddNewtonsoftJson()` it started working but the error persisted because they weren't in their own controller file/

I ended up refactoring my calls into a controller file and that, as expected, silenced the error and I was able to write to my supabase instance.

Since the outcome of this issue ended up being the same [as this one](https://github.com/supabase-community/supabase-csharp/issues/72) I'll probably close this out unless a maintainer of this repo as additional comments.

**Comment by wiverson** [comment]
FWIW a few thoughts: Unity is still on Newtonsoft (a custom fork, no less). So switching from Newtonsoft to System.Text.Json would cut off Unity. The rationale would be that anyone can use the client as is w/Newtonsoft but it cuts out Unity if you move off.

I did a bunch of work a bit ago to get the auth stuff to work with Unity. I only really touched auth to get a bunch of the client-side specific stuff working, but I only use Supabase rpc calls via PostgREST. I don't use any of the pseudo-ORM/GraphGL/dynamic client stuff on either the client or the server.

The projects that I was working on w/Unity all moved to either TypeScript web projects or Godot, so I haven't touched Unity or Supabase C# for a while. So, even though I did a bunch of commits a bit ago I don't really use anything in Supabase C# nowadays.

FWIW for most C# on the server usages I think you would be just fine using a database driver to talk to Postgres directly and/or use a proper C# ORM. You can set up the JS client to manage identity and include auth tokens from web views to REST calls back to your C# REST services. For most of my smaller projects I just use PostgREST and don't even need a middle tier.

@acupofjose might chime in with some thoughts on the status of the repo going forward, I have no idea the current status/maintainers.

**Comment by acupofjose** [comment]
Appreciate the detailed issue and fix @jimweigandt and @frasermclean!

As to status (@wiverson), I've stepped down from maintaining the Supabase C# libs, different obligations and life changes now. I've made the Supabase team aware of that, but AFAIK there hasn't been another collaborator/contributor to step in and fill the need. So, right now, the library is in an unmaintained state.

Agreed that the shift needs to happen to `System.Text.Json`, but I can't be the one to make that happen. At the time this was written (several years ago now) Newtonsoft was the go-to and it's.... not so much anymore.

Anyway, I know that Supabase would be open to new collaborators and they have been willing to pay in the past those that continue to maintain their libraries. So if either of you are interested, feel free to reach out to them!

Apologies that I can't be of more help!
